File: python/meteroid/models/invoice.py
# this file is @generated
import dataclasses
import typing as t
from datetime import datetime

from ..serialization import BaseModel
from .coupon_line_item import CouponLineItem
from .currency import Currency
from .customer_details import CustomerDetails
from .customer_id import CustomerId
from .e_invoicing_status import EInvoicingStatus
from .invoice_id import InvoiceId
from .invoice_line_item import InvoiceLineItem
from .invoice_payment_status import InvoicePaymentStatus
from .invoice_status import InvoiceStatus
from .invoice_type import InvoiceType
from .subscription_id import SubscriptionId
from .tax_breakdown_item import TaxBreakdownItem
from .transaction import Transaction


@dataclasses.dataclass
class Invoice(BaseModel):
    amount_due: int

    applied_credits: int

    coupons: t.List[CouponLineItem]

    created_at: datetime

    currency: Currency

    custom_properties: t.Any
    """User-defined custom property values, keyed by definition `key`."""

    customer_details: CustomerDetails

    customer_id: CustomerId

    id: InvoiceId

    invoice_date: str

    invoice_number: str

    invoice_type: InvoiceType

    line_items: t.List[InvoiceLineItem]

    net_terms: int

    payment_status: InvoicePaymentStatus

    status: InvoiceStatus

    subtotal: int

    subtotal_recurring: int

    tax_amount: int

    tax_breakdown: t.List[TaxBreakdownItem]

    total: int

    transactions: t.List[Transaction]

    billing_period_start: t.Optional[str] = None
    """The period/moment this invoice is about — the subscription period start, or the invoice's
    own date for manual/one-off. Stable and always present, distinct from `invoice_date` (the
    emission date). Shown as "Invoice date"."""

    child_invoice_id: t.Optional[InvoiceId] = None

    due_date: t.Optional[str] = None

    einvoicing_status: t.Optional[EInvoicingStatus] = None

    finalized_at: t.Optional[datetime] = None

    marked_as_uncollectible_at: t.Optional[datetime] = None

    memo: t.Optional[str] = None

    paid_at: t.Optional[datetime] = None

    parent_invoice_id: t.Optional[InvoiceId] = None

    purchase_order: t.Optional[str] = None

    reference: t.Optional[str] = None

    subscription_id: t.Optional[SubscriptionId] = None

    updated_at: t.Optional[datetime] = None

    voided_at: t.Optional[datetime] = None

File: go/invoice.go
// this file is @generated
package meteroid

import "time"

type Invoice struct {
	AmountDue int64 `json:"amount_due"`

	AppliedCredits int64 `json:"applied_credits"`

	// The period/moment this invoice is about — the subscription period start, or the invoice's
	// own date for manual/one-off. Stable and always present, distinct from `invoice_date` (the
	// emission date). Shown as "Invoice date".
	BillingPeriodStart *string `json:"billing_period_start,omitempty"`

	ChildInvoiceId *InvoiceId `json:"child_invoice_id,omitempty"`

	Coupons RequiredSlice[CouponLineItem] `json:"coupons"`

	// RFC 3339 timestamp.
	CreatedAt time.Time `json:"created_at"`

	Currency Currency `json:"currency"`

	// User-defined custom property values, keyed by definition `key`.
	CustomProperties RequiredMap[any] `json:"custom_properties"`

	CustomerDetails CustomerDetails `json:"customer_details"`

	CustomerId CustomerId `json:"customer_id"`

	DueDate *string `json:"due_date,omitempty"`

	// RFC 3339 timestamp.
	FinalizedAt *time.Time `json:"finalized_at,omitempty"`

	Id InvoiceId `json:"id"`

	InvoiceDate string `json:"invoice_date"`

	InvoiceNumber string `json:"invoice_number"`

	InvoiceType InvoiceType `json:"invoice_type"`

	LineItems RequiredSlice[InvoiceLineItem] `json:"line_items"`

	// RFC 3339 timestamp.
	MarkedAsUncollectibleAt *time.Time `json:"marked_as_uncollectible_at,omitempty"`

	Memo *string `json:"memo,omitempty"`

	NetTerms int32 `json:"net_terms"`

	// RFC 3339 timestamp.
	PaidAt *time.Time `json:"paid_at,omitempty"`

	ParentInvoiceId *InvoiceId `json:"parent_invoice_id,omitempty"`

	PaymentStatus InvoicePaymentStatus `json:"payment_status"`

	PurchaseOrder *string `json:"purchase_order,omitempty"`

	Reference *string `json:"reference,omitempty"`

	Status InvoiceStatus `json:"status"`

	SubscriptionId *SubscriptionId `json:"subscription_id,omitempty"`

	Subtotal int64 `json:"subtotal"`

	SubtotalRecurring int64 `json:"subtotal_recurring"`

	TaxAmount int64 `json:"tax_amount"`

	TaxBreakdown RequiredSlice[TaxBreakdownItem] `json:"tax_breakdown"`

	Total int64 `json:"total"`

	Transactions RequiredSlice[Transaction] `json:"transactions"`

	// RFC 3339 timestamp.
	UpdatedAt *time.Time `json:"updated_at,omitempty"`

	// RFC 3339 timestamp.
	VoidedAt *time.Time `json:"voided_at,omitempty"`
}
